nixos/modules/services/networking/nats.nix

```diff
@@ -0,0 +1,174 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.nats;
+
+  confFile = pkgs.writeText "nats.conf" (builtins.toJSON cfg.settings);
+
+in {
+
+  ### Interface
+
+  options = {
+    services.nats = {
+      enable = mkEnableOption "NATS messaging system";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.nats-server;
+        defaultText = "pkgs.nats-server";
+        description = "The nats-server package to use";
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "nats";
+        description = "User account under which NATS runs.";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "nats";
+        description = "Group under which NATS runs.";
+      };
```

Comment:
Why do we need this options?

Reply:
NATS JetStream can use file storage to store messages in it's distributed streams. I thought a user account is needed for the systemd unit in this case, since it needs a stable uid between restarts. Or can this also be done with `DynamicUser`?

Reply:
If you use `StateDirectory` in your `systemd` service it will ensure proper ownership every time. The caveat is that your data directory must reside under `/var/lib`.

Reply:
I've implemented the `StateDirectory` but am wondering what the preferred approach is now: have a user/group created like currently done, or have `DynamicUser = true` and restrict the sysadmin to always put the `dataDir` in `/var/lib`? I'm nudging towards the former because that provides more flexilibity.

Reply:
Dealers choice I guess. We promote `DynamicUser` but sometimes that just doesn't work.

It would be nice if we had a good design pattern in options to specify dynamic use vs NixOS provided system user vs sysadmin provided user. Currently our pattern only adequately covers either the first option **or** the last two options.

Reply:
For simplicity sake, I would prefer `DynamicUser`, however I found another use case for the old-style users: mounting secrets with https://github.com/Mic92/sops-nix In the current implementation of this service that is not yet possible, but it can be extended that way (by having the `configFile` at a fixed location, say `/etc/nats/nats-server.conf` and then symlinking the secret part of the config to the `sops-nix` secret location).

With that in mind I would like to keep the current implementation. If there is a good design pattern to have the sysadmin choose `DynamicUser` over old-style users that would be even better.